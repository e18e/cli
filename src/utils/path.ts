export function normalizePath(path: string): string {
  return path.replaceAll('\\', '/');
}
