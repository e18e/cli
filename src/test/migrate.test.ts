import {describe, it, expect, beforeEach, afterEach} from 'vitest';
import path from 'node:path';
import fs from 'node:fs/promises';
import {
  createTempDir,
  cleanupTempDir,
  runCliProcess,
  stripVersion
} from './utils.js';

describe('migrate command', () => {
  let tempDir: string;

  beforeEach(async () => {
    // Create a temporary directory for each test
    tempDir = await createTempDir();

    // Copy the basic-chalk fixture to the temp dir
    const fixturePath = path.join(process.cwd(), 'test/fixtures/basic-chalk');
    await fs.cp(fixturePath, tempDir, {recursive: true});
  });

  afterEach(async () => {
    // Clean up the temporary directory after each test
    await cleanupTempDir(tempDir);
  });
  it('should migrate with --all flag', async () => {
    const {stdout, stderr, code} = await runCliProcess(
      ['migrate', '--all'],
      tempDir
    );

    expect(code).toBe(0);
    expect(await stripVersion(stdout, tempDir)).toMatchSnapshot();
    expect(stderr).toBe('');

    // Check that the file was actually modified (chalk import should be replaced with picocolors)
    const mainJsPath = path.join(tempDir, 'lib/main.js');
    const fileContent = await fs.readFile(mainJsPath, 'utf-8');
    expect(fileContent).toMatchSnapshot();
  });

  it('should migrate specific package', async () => {
    const {stdout, stderr, code} = await runCliProcess(
      ['migrate', 'chalk'],
      tempDir
    );

    expect(code).toBe(0);
    expect(await stripVersion(stdout, tempDir)).toMatchSnapshot();
    expect(stderr).toBe('');

    // Check that the file was actually modified (chalk import should be replaced with picocolors)
    const mainJsPath = path.join(tempDir, 'lib/main.js');
    const fileContent = await fs.readFile(mainJsPath, 'utf-8');
    expect(fileContent).toMatchSnapshot();
  });

  it('should handle interactive mode', async () => {
    // Test interactive mode by providing input to the prompt
    // Press Enter to accept the default selection
    const {stdout, stderr, code} = await runCliProcess(
      ['migrate', '--all', '--interactive'],
      tempDir,
      '\n' // Press Enter to accept default
    );

    expect(code).toBe(0);
    expect(await stripVersion(stdout, tempDir)).toMatchSnapshot();
    expect(stderr).toBe('');

    // Check that the file was actually modified
    const mainJsPath = path.join(tempDir, 'lib/main.js');
    const fileContent = await fs.readFile(mainJsPath, 'utf-8');
    expect(fileContent).toMatchSnapshot();
  });

  it('should handle custom include pattern', async () => {
    const {stdout, stderr, code} = await runCliProcess(
      ['migrate', 'chalk', '--include', '**/*.js'],
      tempDir
    );

    expect(code).toBe(0);
    expect(await stripVersion(stdout, tempDir)).toMatchSnapshot();
    expect(stderr).toBe('');

    // Check that the file was actually modified
    const mainJsPath = path.join(tempDir, 'lib/main.js');
    const fileContent = await fs.readFile(mainJsPath, 'utf-8');
    expect(fileContent).toMatchSnapshot();
  });

  it('should not modify files with --all flag in dry-run mode', async () => {
    const {stdout, stderr, code} = await runCliProcess(
      ['migrate', '--all', '--dry-run'],
      tempDir
    );

    expect(code).toBe(0);
    expect(await stripVersion(stdout, tempDir)).toMatchSnapshot();
    expect(stderr).toBe('');

    // Check that the file was NOT modified in dry-run mode
    const mainJsPath = path.join(tempDir, 'lib/main.js');
    const fileContent = await fs.readFile(mainJsPath, 'utf-8');
    expect(fileContent).toMatchSnapshot();
  });

  it('should not modify files with specific package in dry-run mode', async () => {
    const {stdout, stderr, code} = await runCliProcess(
      ['migrate', 'chalk', '--dry-run'],
      tempDir
    );

    expect(code).toBe(0);
    expect(await stripVersion(stdout, tempDir)).toMatchSnapshot();
    expect(stderr).toBe('');

    // Check that the file was NOT modified in dry-run mode
    const mainJsPath = path.join(tempDir, 'lib/main.js');
    const fileContent = await fs.readFile(mainJsPath, 'utf-8');
    expect(fileContent).toMatchSnapshot();
  });
});
