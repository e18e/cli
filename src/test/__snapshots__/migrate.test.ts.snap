// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`migrate command > should handle custom include pattern 1`] = `
"[36me18e[39m (cli <version>)

[90mâ”Œ[39m  Migrating packages...
[90mâ”‚[39m
[90mâ”‚[39m  Targets: [2mchalk[22m
[90mâ”‚[39m
[32mâ—‡[39m  {cwd}/lib/main.js...
[90mâ”‚[39m
[90mâ”‚[39m  [2mloading {cwd}/lib/main.js[22m
[2K[1A[2K[G[90mâ”‚[39m  [2mloading {cwd}/lib/main.js[22m
[90mâ”‚[39m  [2mmigrating chalk to picocolors[22m
[2K[1A[2K[1A[2K[G[90mâ”‚[39m  [2mloading {cwd}/lib/main.js[22m
[90mâ”‚[39m  [2mmigrating chalk to picocolors[22m
[90mâ”‚[39m  [2mwriting {cwd}/lib/main.js[22m
[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G[90mâ”‚[39m
[32mâ—†[39m  {cwd}/lib/main.js [2m(1 migrated)[22m
[90mâ”‚[39m
[90mâ””[39m  Migration complete.

"
`;

exports[`migrate command > should handle custom include pattern 2`] = `
"import * as pc from 'picocolors';

console.log(pc.cyan('I am _so_ cyan'));
"
`;

exports[`migrate command > should handle interactive mode 1`] = `
"[36me18e[39m (cli <version>)

[90mâ”Œ[39m  Migrating packages...
[?25l[90mâ”‚[39m
[36mâ—†[39m  Select packages to migrate

[36mâ”‚[39m  [2mSearch:[22m [7m[8m_[28m[27m
[36mâ”‚[39m  [32mâ—¼[39m chalk
[36mâ”‚[39m  [2m[2mâ†‘/â†“[22m[2m to navigate â€¢ [2mSpace:[22m[2m select â€¢ [2mEnter:[22m[2m confirm â€¢ [2mType:[22m[2m to search[22m
[36mâ””[39m"
`;

exports[`migrate command > should handle interactive mode 2`] = `
"import chalk from 'chalk';

console.log(chalk.cyan('I am _so_ cyan'));
"
`;

exports[`migrate command > should migrate specific package 1`] = `
"[36me18e[39m (cli <version>)

[90mâ”Œ[39m  Migrating packages...
[90mâ”‚[39m
[90mâ”‚[39m  Targets: [2mchalk[22m
[90mâ”‚[39m
[32mâ—‡[39m  {cwd}/lib/main.js...
[90mâ”‚[39m
[90mâ”‚[39m  [2mloading {cwd}/lib/main.js[22m
[2K[1A[2K[G[90mâ”‚[39m  [2mloading {cwd}/lib/main.js[22m
[90mâ”‚[39m  [2mmigrating chalk to picocolors[22m
[2K[1A[2K[1A[2K[G[90mâ”‚[39m  [2mloading {cwd}/lib/main.js[22m
[90mâ”‚[39m  [2mmigrating chalk to picocolors[22m
[90mâ”‚[39m  [2mwriting {cwd}/lib/main.js[22m
[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G[90mâ”‚[39m
[32mâ—†[39m  {cwd}/lib/main.js [2m(1 migrated)[22m
[90mâ”‚[39m
[90mâ””[39m  Migration complete.

"
`;

exports[`migrate command > should migrate specific package 2`] = `
"import * as pc from 'picocolors';

console.log(pc.cyan('I am _so_ cyan'));
"
`;

exports[`migrate command > should migrate with --all flag 1`] = `
"[36me18e[39m (cli <version>)

[90mâ”Œ[39m  Migrating packages...
[90mâ”‚[39m
[90mâ”‚[39m  Targets: [2mchalk[22m
[90mâ”‚[39m
[32mâ—‡[39m  {cwd}/lib/main.js...
[90mâ”‚[39m
[90mâ”‚[39m  [2mloading {cwd}/lib/main.js[22m
[2K[1A[2K[G[90mâ”‚[39m  [2mloading {cwd}/lib/main.js[22m
[90mâ”‚[39m  [2mmigrating chalk to picocolors[22m
[2K[1A[2K[1A[2K[G[90mâ”‚[39m  [2mloading {cwd}/lib/main.js[22m
[90mâ”‚[39m  [2mmigrating chalk to picocolors[22m
[90mâ”‚[39m  [2mwriting {cwd}/lib/main.js[22m
[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G[90mâ”‚[39m
[32mâ—†[39m  {cwd}/lib/main.js [2m(1 migrated)[22m
[90mâ”‚[39m
[90mâ””[39m  Migration complete.

"
`;

exports[`migrate command > should migrate with --all flag 2`] = `
"import * as pc from 'picocolors';

console.log(pc.cyan('I am _so_ cyan'));
"
`;

exports[`migrate command > should not modify files with --all flag in dry-run mode 1`] = `
"[36me18e[39m (cli <version>)

[90mâ”Œ[39m  Migrating packages...
[90mâ”‚[39m
[90mâ”‚[39m  Targets: [2mchalk[22m
[90mâ”‚[39m
[32mâ—‡[39m  {cwd}/lib/main.js...
[90mâ”‚[39m
[90mâ”‚[39m  [2mloading {cwd}/lib/main.js[22m
[2K[1A[2K[G[90mâ”‚[39m  [2mloading {cwd}/lib/main.js[22m
[90mâ”‚[39m  [2mmigrating chalk to picocolors[22m
[2K[1A[2K[1A[2K[G[90mâ”‚[39m  [2mloading {cwd}/lib/main.js[22m
[90mâ”‚[39m  [2mmigrating chalk to picocolors[22m
[90mâ”‚[39m  [2mwriting {cwd}/lib/main.js[22m
[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G[90mâ”‚[39m
[32mâ—†[39m  {cwd}/lib/main.js [2m(1 migrated)[22m
[90mâ”‚[39m
[90mâ””[39m  Migration complete.

"
`;

exports[`migrate command > should not modify files with --all flag in dry-run mode 2`] = `
"import chalk from 'chalk';

console.log(chalk.cyan('I am _so_ cyan'));
"
`;

exports[`migrate command > should not modify files with specific package in dry-run mode 1`] = `
"[36me18e[39m (cli <version>)

[90mâ”Œ[39m  Migrating packages...
[90mâ”‚[39m
[90mâ”‚[39m  Targets: [2mchalk[22m
[90mâ”‚[39m
[32mâ—‡[39m  {cwd}/lib/main.js...
[90mâ”‚[39m
[90mâ”‚[39m  [2mloading {cwd}/lib/main.js[22m
[2K[1A[2K[G[90mâ”‚[39m  [2mloading {cwd}/lib/main.js[22m
[90mâ”‚[39m  [2mmigrating chalk to picocolors[22m
[2K[1A[2K[1A[2K[G[90mâ”‚[39m  [2mloading {cwd}/lib/main.js[22m
[90mâ”‚[39m  [2mmigrating chalk to picocolors[22m
[90mâ”‚[39m  [2mwriting {cwd}/lib/main.js[22m
[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G[90mâ”‚[39m
[32mâ—†[39m  {cwd}/lib/main.js [2m(1 migrated)[22m
[90mâ”‚[39m
[90mâ””[39m  Migration complete.

"
`;

exports[`migrate command > should not modify files with specific package in dry-run mode 2`] = `
"import chalk from 'chalk';

console.log(chalk.cyan('I am _so_ cyan'));
"
`;
